The Wettest Update (1.13)

mastercoms opened the pull request:
Glowstone is proud to announce the Update Aquatic (well, Mojang is, we're pretty panicked right now)

### Tasks

#### The big one
- [ ] [1.13 API](https://www.spigotmc.org/threads/bukkit-craftbukkit-spigot-bungeecord-1-13-pre7-development-builds.328883/) compatibility, [Summary](https://www.spigotmc.org/threads/minecraft-1-13-what-to-expect.288714/)
  - [ ] Bukkit [1.13 update](https://hub.spigotmc.org/stash/projects/SPIGOT/repos/bukkit/commits/f8b2086d60942eb2cd7ac25a2a1408cb790c222c)

#### The other big one
- [ ] [1.13 Protocol compatibility](https://wiki.vg/Protocol) (#958)

#### Feature completeness
So. Many. Features. This list is WIP and will only cover major feature additions/changes. If you are a developer, feel free to edit and add more items and detail to this list. For a comprehensive list of changes, refer to [the wiki](https://minecraft.gamepedia.com/1.13).

Most features in this list are non-blocking to this PR getting merged, but they'll be nice to implement. However, there are a few things that have to be implemented or else we wouldn't really be 1.13 (mostly just the stuff in the `General` section below).

**General**
- [ ] Data packs support

Can I interrupt this list for a second to just say how _great_ data packs are going to be for Glowstone? They solve a ton of issues with us not being able to replicate game data from vanilla. Ok, let's carry on.

- [ ] Tags
- [ ] Where did our IDs go? (also known as "the flattening") (#960)

**Blocks**
- [ ] Different air types
- [ ] Stripped logs
- [ ] Stripped wood
- [ ] Turtle Egg
- [ ] New water physics behavior

**Items**
- [ ] Debug stick
- [ ] New recipes _TODO_
  - [ ] Wood block
- [ ] Trident weapon
- [ ] New spawn eggs _TODO_

**World**
- [ ] Coral reefs
- [ ] Icebergs
- [ ] Shipwrecks
- [ ] Underwater caves
- [ ] Underwater ruins
- [ ] Kelp
- [ ] Bubble columns
- [ ] Seagrass
- [ ] Sea Pickle
- [ ] Biomes (https://minecraft.gamepedia.com/Java_Edition_1.13-pre5#Changes)
  - [ ] The End
    - [ ] Small End Islands
    - [ ] End Midlands
    - [ ] End Highlands
    - [ ] End Barrens
  - [ ] Ocean
    - [ ] Warm Ocean
    - [ ] Lukewarm Ocean
    - [ ] Cold Ocean
    - [ ] Deep Warm Ocean
    - [ ] Deep Lukewarm Ocean
    - [ ] Deep Warm Ocean
    - [ ] Deep Cold Ocean
    - [ ] Deep Frozen Ocean
    - [ ] Frozen Ocean
- [ ] Buffet world type

**Status Effects/Potions**
- [ ] Slow falling
- [ ] Turtle master
- [ ] Status: Conduit Power
- [ ] Status: Dolphin's Grace

**Enchantments**
- [ ] Channelling
- [ ] Impaling
- [ ] Loyalty
- [ ] Riptide

**Commands**
- [ ] /bossbar
- [ ] /data
- [ ] /datapack
- [ ] /locate
- [ ] /scoreboard render types hearts and integer added
- [ ] /tp an alias of the new /teleport
- [ ] /time supports setting to noon and midnight

More changes listed here: https://minecraft.gamepedia.com/1.13#Command_format_2

**Advancements**

**Entities**
- [ ] Dolphins
- [ ] Drowned
- [ ] Fish _TODO_
  - [ ] Cod
  - [ ] Salmon
  - [ ] Puffer Fish
  - [ ] Tropical Fish
- [ ] Phantoms
- [ ] Turtles
- [ ] Trident (projectile)

mastercoms committed:
1.13 is done...right?

mastercoms committed:
Merge branch 'dev' into 1.13

aramperes committed:
Replace plugin channel names with namespaces

aramperes committed:
Add OPTCHAT and PARTICLE metadata types

aramperes committed:
Add new metadata entries

aramperes committed:
Arrange new protocol opcodes, add todos for new packets

aramperes committed:
Update existing packets

aki-ks committed:
Implement coordinate-based teleporting (#951)

aramperes committed:
Properly implement Particle metadata type

aramperes committed:
Replace != with \!.equals()

aramperes committed:
Add LoginPluginRequest and LoginPluginResponse packets

mastercoms committed:
Merge branch '1.13' of github.com:GlowstoneMC/Glowstone into 1.13

mastercoms committed:
Update Glowkit to 1.13-R0.1

aramperes committed:
Merge branch '1.13' into wip/1.13/protocol

aramperes committed:
Implement changes to MetadataIndex.NAME_TAG

aramperes committed:
Bump version

aramperes committed:
Merge branch '1.13' into wip/1.13/protocol

aramperes committed:
Fix incorrect switch branch

aramperes committed:
Add stub method to get material from ID

aramperes committed:
Update ChunkSection network serialization

aramperes committed:
Add LongArray NBT Tag

aramperes committed:
Start work on deserializing 1.13 chunk sections

aramperes committed:
Push some work on implementing 1.13 API

aramperes committed:
Start work on Biome builder

aramperes committed:
Optional parameter passing to populator class in builder

aramperes committed:
Improvements to GlowBiomeBuilder

aramperes committed:
Break after reaching valid constructor

aramperes committed:
Skeleton for BlockData access via API

Pr0methean commented on src/main/java/net/glowstone/entity/GlowEntity.java:
```diff
@@ -697,9 +697,16 @@ public void setRawLocation(Location location, boolean fall) {
         if (hasMoved()) {
             if (!fall || type == Material.LADDER // todo: horses are not affected
                     || type == Material.VINE // todo: horses are not affected
```
`this instanceof Horse` (or some subtype) `&&`

aramperes replied:
I'm confused about this todo, does it mean horses should take fall damage even on these blocks, or should be nullified entirely for all horses?

Pr0methean commented on src/main/java/net/glowstone/entity/GlowEntity.java:
```diff
@@ -1487,6 +1494,18 @@ public boolean entityInteract(GlowPlayer player, InteractEntityMessage message)
+    @Override
+    public boolean isPersistent() {
+        // TODO: 1.13
+        throw new UnsupportedOperationException("Not supported yet.");
```
Just add `@Getter @Setter private boolean persistent` as a starting point, and move the TODO there.

aramperes replied:
See the other comment on this

Pr0methean commented on src/main/java/net/glowstone/entity/GlowLivingEntity.java:
```diff
@@ -1243,6 +1224,24 @@ public boolean isHandRaised() {
+    @Override
+    public boolean isSwimming() {
+        // TODO: 1.13
+        throw new UnsupportedOperationException("Not implemented yet.");
```
Add `@Getter @Setter private boolean swimming` as a starting point, and move the TODO there.

aramperes replied:
I'd rather have this error thrown if a plugin uses unimplemented APIs than failing silently. At least then, incompatibility is made clear to the end user.

Pr0methean commented on src/main/java/net/glowstone/entity/objects/GlowArmorStand.java:
```diff
@@ -277,26 +280,31 @@ private EquipmentSlot getEquipType(Material mat) {
             case IRON_HELMET:
```
What about the turtle-shell helmet?

aramperes replied:
We should probably group equipment types anyway (or is that already done?). The turtle-shell can be added to such a group.

Pr0methean replied:
I'd suggest, during the next change to Glowkit, adding the equipment slot as a field of Material with a getter.

aramperes committed:
Merge branch 'wip/1.13/protocol' into 1.13

aramperes committed:
Add turtle helmet to helmet slot

aramperes committed:
Rename some parameters in stub methods, stub GlowTag

aramperes committed:
Implement GlowInventory/GlowPlayerInventory, Material#getEquipmentType

aramperes committed:
Update TerrainObject for 1.13

aramperes committed:
Update all trees and BlockStateDelegate to 1.13

aramperes committed:
Replace instances where GRASS was used instead of GRASS_BLOCK

aramperes committed:
Update biomegrid Biome constants

mastercoms committed:
Target 1.13.1

Pr0methean committed:
Merge branch 'dev' into 1.13

aramperes committed:
[WIP] 1.13 Un-flattening (#960)

mastercoms committed:
Update ForestPopulator to 1.13

mastercoms committed:
Update OrePopulator to 1.13

mastercoms committed:
Update DoublePlantDecorator and some more populators to 1.13

mastercoms committed:
Update BlockStone to 1.13

mastercoms committed:
Update BlockSponge, more double plant stuff, Vex and fix stone

mastercoms committed:
WIP BlockDoublePlant update to 1.13

mastercoms committed:
Double plant type updates, fixes to item system

aramperes committed:
Update Glowkit dependency to Aquatic (1.13)

aramperes committed:
Replace plugin channel names with namespaces

aramperes committed:
Add OPTCHAT and PARTICLE metadata types

aramperes committed:
Add new metadata entries

aramperes committed:
Arrange new protocol opcodes, add todos for new packets

aramperes committed:
Fix test and compilation errors

aramperes committed:
Update existing packets

aramperes committed:
Properly implement Particle metadata type

aramperes committed:
Replace != with \!.equals()

aramperes committed:
Add LoginPluginRequest and LoginPluginResponse packets

aramperes committed:
Temporarily remove footstep particle

aramperes committed:
Update protocol version

aramperes committed:
Fix tests, checkstyle, OPTCHAT for getCustomName

aramperes committed:
Add Glowkit branch to build

aramperes committed:
Glowkit wants git username and email

aramperes committed:
Merge branch 'dev' into 1.13-hack

aramperes committed:
Update Slot serialization to 1.13.2

aramperes committed:
New network Mob IDs and Object IDs

aramperes committed:
Checkstyle issues

aramperes committed:
Make world joinable (empty world temporarily)

aramperes committed:
Ground work for BlockData mapping

aramperes committed:
Better hashing for loaders

aramperes committed:
Add test for mutated state changing IDs

mastercoms committed:
Merge branch 'dev' into 1.13-hack

# Conflicts:
#	src/main/java/net/glowstone/entity/GlowPlayer.java

mastercoms committed:
Merge branch 'dev' into 1.13

# Conflicts:
#	src/main/java/net/glowstone/block/GlowBlock.java
#	src/main/java/net/glowstone/block/blocktype/BlockDoublePlant.java
#	src/main/java/net/glowstone/block/blocktype/BlockMushroom.java
#	src/main/java/net/glowstone/block/blocktype/BlockSapling.java
#	src/main/java/net/glowstone/block/blocktype/BlockStem.java
#	src/main/java/net/glowstone/block/blocktype/BlockTallGrass.java
#	src/main/java/net/glowstone/block/blocktype/BlockType.java
#	src/main/java/net/glowstone/command/minecraft/TeleportCommand.java
#	src/main/java/net/glowstone/command/minecraft/TpCommand.java

aramperes committed:
Merge branch 'dev' into 1.13-hack

Commit:
MC-123880: use correct flag on 1.13+ (#1030)

Pr0methean committed:
Merge branch 'dev' into 1.13, and partial bug fixes

# Conflicts:
#	src/main/java/net/glowstone/GlowWorld.java
#	src/main/java/net/glowstone/block/blocktype/BlockSponge.java
#	src/main/java/net/glowstone/command/minecraft/TeleportCommand.java
#	src/main/java/net/glowstone/entity/GlowEntity.java

Pr0methean committed:
Partial update to 1.13.2 for biomes etc.

Pr0methean committed:
More updated biome names, and restore deleted BiomeHeight class

Pr0methean committed:
Update remaining biome names

Pr0methean committed:
Stub out some missing new methods

Pr0methean committed:
Update BlockTallGrass and DiggingHandler

Pr0methean committed:
Update Lombok plugin
Was giving StackOverflowError due to an old bug.

Pr0methean committed:
Implement KeyedBossBar

Pr0methean committed:
More updates

Pr0methean committed:
Updates to GlowPlayer, StatisticMap, ChunkManager

Pr0methean committed:
Use getMaterialFromBaseId where still needed

Pr0methean committed:
Update GlowBlock

Pr0methean committed:
More updates, but still more needed

Pr0methean committed:
More updates

Pr0methean committed:
Fix GlowMetaItem, and set some necessary args in pom.xml

Pr0methean committed:
Update ItemTable, and add necessary TODOs

Pr0methean committed:
Update tool names

Pr0methean committed:
Handle air variants

Pr0methean committed:
Add sanity checks for reflectively-generated sets in MaterialUtil

Pr0methean committed:
Update GlowDesertTemple and GlowBannerMatcher

Pr0methean committed:
Update GlowNoteBlock

Pr0methean committed:
Update FlowerType

Pr0methean committed:
Update Lootable implementation in GlowContainer

Pr0methean committed:
Strengthen MaterialUtil sanity check

Pr0methean committed:
Typo fix

Pr0methean committed:
Update GlowBoat

WesleyVanNeck commented:
Rip 1.14.1 is out

Pr0methean committed:
Merge branch '1.13' into 1.13-hack

Pr0methean committed:
Switch to a Glowkit version that exists in the repo

mastercoms committed:
Merge branch 'dev' into 1.13-hack

# Conflicts:
#	src/main/java/net/glowstone/util/InventoryUtil.java

mastercoms committed:
update glowkit to latest from source repo

mastercoms committed:
starting on 1.13 progress again

bad merge/compiler error fixes
mostly 1.13 inventory progress

mastercoms committed:
fix bad import styling

mastercoms committed:
1.13 tags work, finish up inventory errors

mastercoms committed:
update lombok

mastercoms committed:
fix lombok builds

mastercoms committed:
use stacksize increase on azure

mastercoms committed:
still compile from main sources

mastercoms committed:
more 1.13 work

mastercoms committed:
a bit more 1.13 work

mastercoms committed:
in progress particle system rewrite

mastercoms committed:
finish up new particle representation

mastercoms committed:
clean up unused arrays in new particle class

mastercoms committed:
generator 1.13 progress, breeding api

mastercoms committed:
zombie stubs

mastercoms committed:
fixed all but one compilation error

mastercoms committed:
fix generated flattening data

mastercoms committed:
fix checkstyle errors

mastercoms committed:
fix test compile errors

mastercoms committed:
fix test typo

mastercoms committed:
fix fishing rewards partially

mastercoms committed:
Merge remote-tracking branch 'origin/dev' into 1.13

# Conflicts:
#	src/main/java/net/glowstone/entity/GlowLivingEntity.java
#	src/main/java/net/glowstone/net/http/HttpClient.java
#	src/main/java/net/glowstone/net/protocol/LoginProtocol.java
#	src/test/java/net/glowstone/net/LoginProtocolTest.java

mastercoms committed:
Merge branch 'dev' into 1.13

mastercoms committed:
fix unused imports

mastercoms committed:
fix enchantments

mastercoms committed:
comment up tests

mastercoms committed:
update material values

mastercoms committed:
provide unsafe access in server test class

aramperes committed:
Properly implement OfflinePlayer#getLastLogin (Paper API)

aramperes committed:
Save default block state id and convert from Material

AlexCodesGames committed:
[BlockBed.java] Updated Bed Break Logic (#1059)

Beds now break if the blocks directly below the head or foot bed blocks are changed to air (broken).

mastercoms committed:
Merge branch 'dev' into 1.13

# Conflicts:
#	src/main/java/net/glowstone/constants/GlowBiomeClimate.java
#	src/main/java/net/glowstone/entity/GlowLivingEntity.java

mastercoms committed:
add default ids, use block data for redstone power

mastercoms committed:
add default utilities

mastercoms committed:
use new default IDs for property map in block data

Postremus committed:
Only run delombok in ci builds (#1068)

The source files by delombok are only required for javadoc.
We now hide delombok, javadoc jar, and source jar generation behind profile build-extras.
The profile can be activated on locally as well. This should make it easier to debug ci problems.

PayDevD committed:
Use synchronizedSet (#1072)

Muqsit committed:
Fix wrong coordinates passed to isSnowy (#1075)

mio-19 committed:
github wiki was replaced by Glowstone docs (#1083)

DenyCZ committed:
Add missing options to time command (#1067)

mastercoms committed:
update maven options

Disable Maven launching with advanced JIT optimization for its JVM, makes maven start faster
Skip test compilation because 1.13 tests aren't working yet

mastercoms committed:
update dependencies and use ECJ for compilation

mastercoms commented:
Superseded by #1091
